### ISSUE TITLE
Datasource plugin: CircularDataFrame makes graphs not show up in query editor

### ISSUE THREAD

**Issue by SeresHotes** [grafana/grafana#80139]
### What happened?

I've started developing a datasource plugin for Grafana. I tried to replicate [datasource-streaming-websocket example](https://github.com/grafana/grafana-plugin-examples/blob/main/examples/datasource-streaming-websocket/streaming-websocket-plugin/src/DataSource.ts). I've encountered a problem, that using `CircularDataFrame` makes graphs not show up in datasource query editor. Here's a minimal example.
![image](https://github.com/grafana/grafana/assets/5305710/e79c691e-d672-4c76-9145-217a7393245d)

src/datasource.ts
```TypeScript
import {
  DataQueryRequest,
  DataQueryResponse,
  DataSourceApi,
  DataSourceInstanceSettings,
  FieldType,
  CircularDataFrame,
} from '@grafana/data';
import { LoadingState } from '@grafana/schema';

import { MyQuery, MyDataSourceOptions } from './types';

export class DataSource extends DataSourceApi<MyQuery, MyDataSourceOptions> {
  constructor(instanceSettings: DataSourceInstanceSettings<MyDataSourceOptions>) {
    super(instanceSettings);
  }

  async query(options: DataQueryRequest<MyQuery>): Promise<DataQueryResponse> {
    console.log("query")
    const { range } = options;
    const from = range!.from.valueOf();
    const to = range!.to.valueOf();

    // Return a constant for each query.
    const data = options.targets.map((target) => {
      let frame = new CircularDataFrame({
        append: 'tail',
        capacity: 2,
      })
      frame.addField({ name: 'time', type: FieldType.time });
      frame.addField({ name: 'value', type: FieldType.number });
      frame.add({ 'time': from, 'value': 1})
      frame.add({ 'time': to, 'value': 2})
      return frame;
    });

    return {data, state: LoadingState.Done } ;
  }

  async testDatasource() {
    // Implement a health check for your data source.
    return {
      status: 'success',
      message: 'Success',
    };
  }
}
```

### What did you expect to happen?

Replacing `CircularDataFrame` with `MutableDataFrame` works perfectly. But it doesn't seem desirable for a streaming data.
![image](https://github.com/grafana/grafana/assets/5305710/3ecb74e3-3a43-4a9d-a81d-e81e6b19f39c)
```TypeScript
...
    const data = options.targets.map((target) => {
      let frame = new MutableDataFrame()
      frame.addField({ name: 'time', type: FieldType.time });
      frame.addField({ name: 'value', type: FieldType.number });
      frame.add({ 'time': from, 'value': 1})
      frame.add({ 'time': to, 'value': 2})
      return frame;
    });
...
```

### Did this work before?

Can't test it.

### How do we reproduce it?

1. Create a datasource plugin using `npx @grafana/create-plugin@latest`
2. Replace code in src/datasource.ts with the provided example
3. `npm install`, `npm run dev`, `docker-compose up`
4. Create a new datasource using the plugin
5. Create a new dashboard using the datasource

### Environment (with versions)?

Grafana: Docker image: grafana/grafana-enterprise:10.0.3, 10.0.3-amd64, latest, main
OS: Windows 10 Pro 22H2
Browser: Google Chrome Version 120.0.6099.130 (Official Build) (64-bit)

### Grafana platform?

Docker

**Comment by jackw** [comment]
I'm able to reproduce this but so far I've only understood something changed between Grafana 9 and 10. The [plugin example](https://github.com/grafana/grafana-plugin-examples/blob/main/examples/datasource-streaming-websocket/streaming-websocket-plugin/src/DataSource.ts) streams correctly into the table panel in Grafana 10 so wondering if @grafana/dataviz-squad would be able to help figure out what's changed / if this is a bug in Timeseries panel?

**Comment by baldm0mma** [comment]
@SeresHotes can you confirm that this versioning you're using is correct? `Grafana: Docker image: grafana/grafana-enterprise:10.0.3, 10.0.3-amd64, latest, main` did you mean `10.3.3` or `10.0.3`?

**Comment by SeresHotes** [comment]
@baldm0mma yes, the version was 10.0.3. I didn't try younger or older versions

**Comment by baldm0mma** [comment]
I looked more into this, and it appears that `CircularDataFrame` is pretty [gutted](https://github.com/grafana/grafana/blob/main/packages/grafana-data/src/dataframe/CircularDataFrame.ts) (both `CircularVector` and `MutableDataFrame` are both deprecated). Looks like we now use `StreamingDataFrame` as a replacement for the data streaming piece. @jackw I think the example might just need to be updated? I was able to get data visualized with the following included `StreamingDataFrame` changes to the example (disclaimer: I've never written a plugin before, maybe this change is terrible!). 😅

```
import {
  DataQueryRequest,
  DataQueryResponse,
  DataSourceApi,
  DataSourceInstanceSettings,
  FieldType,
  StreamingDataFrame,
} from '@grafana/data';
import { LoadingState } from '@grafana/schema';

import { MyQuery, MyDataSourceOptions } from './types';

export class DataSource extends DataSourceApi<MyQuery, MyDataSourceOptions> {
  constructor(instanceSettings: DataSourceInstanceSettings<MyDataSourceOptions>) {
    super(instanceSettings);
  }

  async query(options: DataQueryRequest<MyQuery>): Promise<DataQueryResponse> {
    console.log('query');
    const data = options.targets.map((_target) => {
      const json = {
        schema: {
          fields: [
            { name: 'time', type: FieldType.time },
            { name: 'name', type: FieldType.string },
            { name: 'value', type: FieldType.number },
          ],
        },
        data: {
          values: [
            [100, 200, 300],
            ['a', 'b', 'c'],
            [1, 2, 3],
          ],
        },
      };

      return StreamingDataFrame.fromDataFrameJSON(json, {
        maxLength: 5,
        maxDelta: 300,
      });
    });

    return { data, state: LoadingState.Done };
  }

  async testDatasource() {
    // Implement a health check for your data source.
    return {
      status: 'success',
      message: 'Success',
    };
  }
}
```
<img width="1440" alt="Screenshot 2024-02-29 at 12 55 48 PM" src="https://github.com/grafana/grafana/assets/46619047/20ded097-99cc-4270-aede-7043af9fd23c">

Also, @SeresHotes there's a lot of great updates in the most recent Grafana version, and I would advise writing your plugin in 10.3 or later.

**Comment by leeoniya** [comment]
yep, `CircularDataFrame` will be going away at some point, all of its internals are deprecated already:

![image](https://github.com/grafana/grafana/assets/43234/34812ba4-ac99-49bf-9958-493e3f9f1349)

---

Long live `StreamingDataFrame`:

https://github.com/grafana/grafana/blob/42b55aedbc3e35ed426ff684347492cbdbea3314/packages/grafana-data/src/dataframe/StreamingDataFrame.ts#L74

we should update the remaining uses of our core internals to use `StreamingDataFrame`, and update [datasource-streaming-websocket example](https://github.com/grafana/grafana-plugin-examples/blob/main/examples/datasource-streaming-websocket/streaming-websocket-plugin/src/DataSource.ts) as well.
